### ISSUE TITLE
Change Gson to faster serialization library

### ISSUE THREAD

**Issue by andreasdc** [SpigotMC/BungeeCord#3624]
### Feature description

I was looking and the profiler and saw that serializing and deserializing of strings such as scoreboard components is taking some processing time, maybe it would be a good idea to change it to a faster library? The servers have scoreboard plugin, I think tablist plugin like https://github.com/CodeCrafter47/BungeeTabListPlus would have better performance too?
![image](https://github.com/SpigotMC/BungeeCord/assets/82218780/9c1af53b-48af-44d5-90e1-c6cea3f8c2fd)
![image](https://github.com/SpigotMC/BungeeCord/assets/82218780/7d70c9c3-9897-42da-ac24-e619914f73c7)

### Goal of the feature

When looking at the results from here, gson, which is currently being used in BungeeCord, got 8677 ns, while for example fury-fastest got 442 and jackson 1994. This is almost 20x improvement with fury or 4x with jackson.
https://github.com/eishay/jvm-serializers/wiki

### Unfitting alternatives

-

### Checking

- [X] This is not a question or plugin creation help request.
- [X] This is a **feature or improvement request**.
- [ ] I have not read these checkboxes and therefore I just ticked them all.
- [X] I did not use this form to report a bug.

**Comment by Outfluencer** [comment]
i personally added a componen serial and deserialisation cache to my bungee do speed exactly this up

**Comment by andreasdc** [comment]
> i personally add a component serial and deserialisation cache to my bungee do speed exactly this up I think not using gson is a bad idea, as gson is the most widely used lib

Can I see how did you do that? Did you see any improvement? If the behaviour will be exactly the same and only the performance will be better, I think that's a good idea to implement that.

**Comment by Outfluencer** [comment]
Behaviour is the same but i have no benchmarks

**Comment by Outfluencer** [comment]
I added a threadlocal limited cache that only caches the most used components

**Comment by andreasdc** [comment]
> I added a threadlocal limited cache that only caches the most used components

With many components that don't repeat it will be slower than normal one probably.

**Comment by Outfluencer** [comment]
In most cases one component will be sent to multiple players and gets deserized multiple times so in most cases its much faster

**Comment by Outfluencer** [comment]
The only case its, i would not even call ist slower, but has no performance boost is if you senden one single component to only one player

**Comment by andreasdc** [comment]
Oh yes with some players I see that it will improve by a bit, even proxy pinging will be improved, along with chat, bossbar, scoreboard and yes, every tab component like Item, Header/Footer uses gson too (BungeeTabListPlus would be improved already, because it uses ComponentSerializer.deserialize(). If it won't break anything and would help up to 20x, then I think it's a good idea.
![image](https://github.com/SpigotMC/BungeeCord/assets/82218780/559cf401-0455-4cae-95c1-51f04e6051a2)
![image](https://github.com/SpigotMC/BungeeCord/assets/82218780/b4756d3d-2283-4133-824d-7d37b7cbbd10)
![image](https://github.com/SpigotMC/BungeeCord/assets/82218780/6596b727-677c-4d86-96fd-7d65d046ecd2)

**Comment by Janmm14** [comment]
Just because a library is widely used doesn't mean that bungee should stick to it. Bungee would only have to carry the library as dead weight around in case plugins use it for a couple years (file size is not a bug prolem)

**Comment by andreasdc** [comment]
> Just because a library is widely used doesn't mean that bungee should stick to it. Bungee would only have to carry the library as dead weight around in case plugins use it for a couple years (file size is not a big problem), but it could switch to a faster serialization library at an time, **especially when a new library seems to be an order of magnitude faster in benchmarks**.
> 
> Edit: However it seems like Fury for example is not yet in a usable state. It does not contain json (de-)serialization.

That's not it?
![image](https://github.com/SpigotMC/BungeeCord/assets/82218780/3b676a66-f8b9-443a-b97f-8faf9917c271)

**Comment by Janmm14** [comment]
> > Just because a library is widely used doesn't mean that bungee should stick to it. Bungee would only have to carry the library as dead weight around in case plugins use it for a couple years (file size is not a big problem), but it could switch to a faster serialization library at an time, **especially when a new library seems to be an order of magnitude faster in benchmarks**.
> > Edit: However it seems like Fury for example is not yet in a usable state. It does not contain json (de-)serialization.
> 
> That's not it? ![image](https://private-user-images.githubusercontent.com/82218780/308286021-3b676a66-f8b9-443a-b97f-8faf9917c271.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNTcxNTAsIm5iZiI6MTcwOTA1Njg1MCwicGF0aCI6Ii84MjIxODc4MC8zMDgyODYwMjEtM2I2NzZhNjYtZjhiOS00NDNhLWI5N2YtOGZhZjk5MTdjMjcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDE4MDA1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhMTE1YzhlNDY2M2RkYmE1OTkwZjc5MmYwMjRmZjU3YTk4NzQ0NmExNzdhZmYwMmIzNDg2NjFlMmRmNDY0ODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Kd9zW_dxlvWvOtfRRibfJpKcXE-dZxL60zrsn8SbO0I)

No, that does not create json, it serializes to bytes with a custom format or maybe its using jvm's serialization format.
I am not sure whether Fury is actually intended to serialize to json at all, given there is no mention of such json in explixitely in documentation or code.
Also the existing documentation is lacking a lot.

**Comment by andreasdc** [comment]
Hmm, the fastest that has json in it's name is: https://github.com/ngs-doo/dsl-json - which is almost 8x faster than gson.

**Comment by Outfluencer** [comment]
problem with cache is that a cached component is mutable, in most cases thats not rpoblem but if you use components weirdly it can be

**Comment by Outfluencer** [comment]
that i i thought we could do it

![image](https://github.com/SpigotMC/BungeeCord/assets/48880402/2da95bd7-de62-4dc7-be32-153c1aebc8fa)

**Comment by Outfluencer** [comment]
in this case IdentityLinkedHashMap is limited to 128 and the oldest (the oldest one is the one that is used the least) value in der cache is removed any time a new is added

**Comment by andreasdc** [comment]
I don't think using cache, when you don't know what will go in the json, is a good idea. Many json strings will be unique, having different things like stats, names, ping etc. You could technically check if it's unique, but it will still use extra resources. IMO using better library will increase json's performance by up to 8x and should bring great result. 🚀 💘

**Comment by Outfluencer** [comment]
> I don't think using cache, when you don't know what will go in the json, is a good idea. Many json strings will be unique, having different things like stats, names, ping etc. You could technically check if it's unique, but it will still use extra resources. IMO using better library will increase json's performance by up to 8x and should bring great result. 🚀 💘

thats why i am checking for identity instead so we dont lose any performance

**Comment by Janmm14** [comment]
> > I don't think using cache, when you don't know what will go in the json, is a good idea. Many json strings will be unique, having different things like stats, names, ping etc. You could technically check if it's unique, but it will still use extra resources. IMO using better library will increase json's performance by up to 8x and should bring great result. 🚀 💘
> 
> thats why i am checking for identity instead so we dont lose any performance

This is no suitable general solution as BaseComponents are not final and itself or its extra's could change at any point in time without identity changing.

With an identity based hash map, I do not see any actual cache hits in vanilla bungeecord operation.

**Comment by andreasdc** [comment]
> > > > I don't think using cache, when you don't know what will go in the json, is a good idea. Many json strings will be unique, having different things like stats, names, ping etc. You could technically check if it's unique, but it will still use extra resources. IMO using better library will increase json's performance by up to 8x and should bring great result. 🚀 💘
> > > 
> > > 
> > > thats why i am checking for identity instead so we dont lose any performance
> > 
> > 
> > This is no suitable general solution as BaseComponents are not final and itself or its extra's could change at any point in time without identity changing.
> > With an identity based hash map, I do not see any actual cache hits in vanilla bungeecord operation.
> 
> you're 100% right its not any solution i would recommend or make a pr for but maybe in this case it could help him
> 
> In my opinion we should not change lib

What's the reason you prefer gson?

**Comment by Outfluencer** [comment]
most used json lib ig, also i personally used it in all of my projects, i dont think we should recode all json related code just to get 0,01% better cpu times, also components are not sent by the client so its not exploitable

**Comment by Janmm14** [comment]
I took a look at some of the java json libraries ranking high here: https://github.com/fabienrenaud/java-json-benchmark
I looked at fastjson, dsljson and avajejsonb.

Fastjson is missing documentation in english. The other two are also kinda missing some documentation for the more customizable parts.

I didnt look further into fastjson, but dsljson and avajejsonb are missing the flexibility of redirecting the target deserialization object like we do in ComponentSerializer class.

I did not look at more libraries, the performance improvement is not thathigh anymore.
I also think that with bungee's heavy use of custom serialization we are likely near the top of gson's performance as there is less reflection involved.
(I did not look at the gson usage in the benchmark, but if it doesn't use registered serializers but more reflection instead, there's a chance we are even faster than the bench measured.)

**Comment by andreasdc** [comment]
I think it would help massively, I summarized all gson usage and it was 4,12%, it will help with scoreboard, bossbar, chat, server pinging, BungeeTabListPlus and some more. It was calculated during a big attack, that had no impact on gson. Without the attack it looks like gson operations are taking 8% of cpu time.

**Comment by Outfluencer** [comment]
And what do you expect from the change? 1% less cpu times?

**Comment by andreasdc** [comment]
> And what do you expect from the change? 1% less cpu times?

Well initially I saw 8x improvement, but idk which lib is possible to use.

**Comment by bob7l** [comment]
Bungee's serialization is complicated and deep. Changing the json library probably wouldn't net a significant performance improvement. Deeper profiling should be done to see exactly what's taking the most time. If I got time I'll drop some timings from Yourkit  here

**Comment by andreasdc** [comment]
I did my timings using Spark and I described the usage above.

**Comment by bob7l** [comment]
> I did my timings using Spark and I described the usage above.

Yeah, but profiling with something like YourKit will show very detailed timings. 

Anyways, here's some profiled results. This is from 130,584 total milliseconds. Just looked into deserialization since it's likely the bigger resource hog. 

Part1: https://i.gyazo.com/9d56a65540433521f7e34b7b33943137.png
Part2: https://i.gyazo.com/e0fa9bf1dc0141feb981bca81a1e7e2c.png

Flamegraph: https://i.gyazo.com/d5c470b668985eccb58b325dc67e8ce7.png

This is just on a test server with a bunch of fake (1.8.9) players and a single real 1.20.4 player, so the data isn't too good. Would be interesting to see with live players. 

A large amount of the timings are just from Bungee's deserialization. Which is to be expected, it's pretty intensive. A good amount of timings also go to  Gson's internal usage of LinkedTreeMap. I imagine most if not all of these libraries use a linked hashmap or a similar implementation with string keys. 


Here's one part of the flamegraph I found interesting: https://i.gyazo.com/3e392e51e317ce3886c55f5dfda8ac0e.png

The toUppercase is using a pretty large amount of time: https://github.com/SpigotMC/BungeeCord/blob/1b88a8471077929bcfbd3a5bd6c7cfdf93df92de/chat/src/main/java/net/md_5/bungee/api/ChatColor.java#L267C49-L267C60

Which would likely be a pretty simple optimization.

**Comment by bob7l** [comment]
> > Which would likely be a pretty simple optimization.
> 
> What's the optimisation?

I meant the toUppercase could be easily optimized, not the whole thing. Whole thing is already well written. 

For the uppercase, you could add the different uppercase/lowercase variants to the map rather then changing the string case. Would be a couple thousand entries in total though.

It's only 8.33% of the ChatColor.of function though so mostly a micro optimization...

**Comment by Outfluencer** [comment]
Thats sounds like a terrible idea

**Comment by bob7l** [comment]
> Thats sounds like a terrible idea

It's not really worth it, but it's not a terrible idea. Very very common practice to achieve O(1). And realistically a tiny memory footprint

**Comment by andreasdc** [comment]
> > I did my timings using Spark and I described the usage above.
> 
> Yeah, but profiling with something like YourKit will show very detailed timings.
> 
> Anyways, here's some profiled results. This is from 130,584 total milliseconds. Just looked into deserialization since it's likely the bigger resource hog.
> 
> Part1: https://i.gyazo.com/9d56a65540433521f7e34b7b33943137.png Part2: https://i.gyazo.com/e0fa9bf1dc0141feb981bca81a1e7e2c.png
> 
> Flamegraph: https://i.gyazo.com/d5c470b668985eccb58b325dc67e8ce7.png
> 
> This is just on a test server with a bunch of fake (1.8.9) players and a single real 1.20.4 player, so the data isn't too good. Would be interesting to see with live 1.20.4+ players.
> 
> A large amount of the timings are just from Bungee's deserialization. Which is to be expected, it's pretty intensive. A good amount of timings also go to Gson's internal usage of LinkedTreeMap. I imagine most if not all of these libraries use a linked hashmap or a similar implementation with string keys.
> 
> Here's one part of the flamegraph I found interesting: https://i.gyazo.com/3e392e51e317ce3886c55f5dfda8ac0e.png
> 
> The toUppercase is using a pretty large amount of time: https://github.com/SpigotMC/BungeeCord/blob/1b88a8471077929bcfbd3a5bd6c7cfdf93df92de/chat/src/main/java/net/md_5/bungee/api/ChatColor.java#L267C49-L267C60
> 
> Which would likely be a pretty simple optimization.
> 
> Second minor optimization in that same function would be to just flip `( string.startsWith( "#" ) && string.length() == 7 )` so it checks the length first

How did you benchmark it? I think serialization is mostly done on higher versions.

**Comment by bob7l** [comment]
> No adding only all combinations for strikethrough would be be 8192 string entrys only for one single string now multiply that by the amount of possible other colors/styles and you would notice its not a small memory footprint

There would be a total of 20,456 different strings. Might be 1-2MB in total, if even that. 

Could also just key by the hashcodes assuming there's no collisions between those 20,456 strings, but that could lead to unintended behavior if there's a collision with a random unknown that was given for color. 

Again though, this is based on being asked an optimization and responding. I wouldn't bother with this as optimizing toUpperCase would just be a tiny chip - or in other words micro-optimization. Not to mention hex colors, which have quickly become adopted, are the new norm.

**Comment by Outfluencer** [comment]
if i look into spark profiler i cant see any extensive gson calls

**Comment by andreasdc** [comment]
> if i look into spark profiler i cant see any extensive gson calls

They need to be actually used

**Comment by Outfluencer** [comment]
its used in the networking isn't it?

**Comment by bob7l** [comment]
> 
> How did you benchmark it? I think serialization is mostly done on higher versions and you see it heavily used with tablist and scoreboard.
Everything is deserialized into a BaseComponent now for simplicity.

**Comment by Outfluencer** [comment]
lazyloading would be implementable for all packets that are directly try to parse it as an basecomponent

**Comment by Janmm14** [comment]
Its a little contradicting that we have BY_NAME map using upper case while we serialize to lower case.
We should examine whether other chat libraries are using uppercase or lowercase after serializartion and then switch our serialization or BY_NAME map. My guess for this is that it might be faster to call lowercase on something that is already lowercase. In theory we could also attempt using length and charAt checks with a final equals check against single values instead of a map for optimization. We do not expect any more named colors and some new boolean style option is also not expected.

What I noticed more in the benchmark is the huge JsonElement.has usage inside CompoenentStyleSerializer, actually taking most of the deserialization time.
This can be a little optimized by using get and a null check to optimize "hits".
Assuming that we nearly never get chat json with illegal content and that often less than 3 of the 7 style options will be set, it should be worth switching to iterating over the EntrySet of the json object and using a switch inside.

I have pushed these little optimizations of the 2nd paragraph here: https://github.com/Janmm14/BungeeCord/tree/chat-optimization
@bob7l can you compare the performance maybe?

**Comment by bob7l** [comment]
> I don't see JsonElement.has in my profilers. ![image](https://private-user-images.githubusercontent.com/82218780/311289588-28254701-11f3-47f5-b9ad-d5e3633d25bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk5MTg1NzIsIm5iZiI6MTcwOTkxODI3MiwicGF0aCI6Ii84MjIxODc4MC8zMTEyODk1ODgtMjgyNTQ3MDEtMTFmMy00N2Y1LWI5YWQtZDVlMzYzM2QyNWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA4VDE3MTc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5Yzg5NDY5YzU5YzU1YTQ2MjA0ZTBjOWM4NDdjMjA3ZGFiYTcxNjY5NGZhZjg3ZmIwYzViYWIxNzA4MTk4NjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.z3yU0leIqn-PNdRrPBK-Ch37El7eZjCavjvv4hNlN7E)

Serialization/Deserialization differences aside, you're using Spark's sampler. While it's good for getting a ballpark idea of what's taking up time, it's not very accurate and usually wont show quick functions like map.get and so on. 

Whereas I'm using instrumentation/tracing which offers a more precise measurement and is able to pickup on finer details. It's also a lot more useful in these cases with massive trees from having to iterate over many children components. 

You should be able to grab a free trial of YourKit. Incredible profiler

**Comment by andreasdc** [comment]
ComponentSerializer.toString uses gson.toJson, there's nothing about JsonElement.has.

**Comment by Janmm14** [comment]
> ComponentSerializer.toString uses gson.toJson, there's nothing about JsonElement.has.

Yes. That is the "serializing" part: java objects to json string.

gson is also involved in the "deserializing" part: json string to java objects.
This is done with ComponentSerializer.deserialize method.

Different method, still being used a lot in bungee.

If we go to handling "chat" data from packets only on demand, that could save us both steps in a lot of cases, leaving only bungee plugins sending chat to be affected on the bungee side (as the chat part is also used in spigot, serialization speed is still a concern somewhat tho)

**Comment by Janmm14** [comment]
In branch [chat-optimization-v2](https://github.com/Janmm14/BungeeCord/tree/chat-optimization-v2) I have created an implementation of lazy deserialization in packets with methods to hopefully retain bytecode compatibility.

**Comment by andreasdc** [comment]
> In branch [chat-optimization-v2](https://github.com/Janmm14/BungeeCord/tree/chat-optimization-v2) I have created an implementation of lazy deserialization in packets, with methods to hopefully retain full bytecode compatibility (Edit: for the individual packets, not for DefinedPacket methods).
> 
> Edit 2: Its untested btw. Maybe someone has a better way of implementing it? It should be possible to replace the long generics of Deserializable with static types, for shorter code, but I'm unsure if thats desired over flexibility towards possible similar attempts for other complex deserialization.

One of the patches https://github.com/SpigotMC/BungeeCord/pull/3629/commits/dd338c5fd546f31c7b459871cf73c425b43adc55 this or that https://github.com/Janmm14/BungeeCord/commit/5b12adb97f0e9db28c123a4d3ef88fd0909ed730 I think broke bold options from the tablist's entries.

**Comment by andreasdc** [comment]
And it's the same for strikethrough too.

**Comment by Janmm14** [comment]
weird, i will test and investigate and find the probably dumb mistake i made

**Comment by andreasdc** [comment]
Weird even more, because I'm testing with master and I still have this issue.

**Comment by andreasdc** [comment]
Yea, that's what I thought happened, it's an issue inside the BungeeCord. https://github.com/SpigotMC/BungeeCord/issues/3631

**Comment by bob7l** [comment]
> In branch [chat-optimization-v2](https://github.com/Janmm14/BungeeCord/tree/chat-optimization-v2) I have created an implementation of lazy deserialization in packets, with methods to hopefully retain full bytecode compatibility (Edit: for the individual packets, not for DefinedPacket methods).
> 
> Edit 2: Its untested btw. Maybe someone has a better way of implementing it? It should be possible to replace the long generics of Deserializable with static types, for shorter code, but I'm unsure if thats desired over flexibility towards possible similar attempts for other complex deserialization.

Probably the most basic way of implementing it without breaking anything. Hopefully has a chance of getting merged.. De/serialization is only going to get heavier and heavier in the future.

**Comment by andreasdc** [comment]
> > In branch [chat-optimization-v2](https://github.com/Janmm14/BungeeCord/tree/chat-optimization-v2) I have created an implementation of lazy deserialization in packets, with methods to hopefully retain full bytecode compatibility (Edit: for the individual packets, not for DefinedPacket methods).
> > Edit 2: Its untested btw. Maybe someone has a better way of implementing it? It should be possible to replace the long generics of Deserializable with static types, for shorter code, but I'm unsure if thats desired over flexibility towards possible similar attempts for other complex deserialization.
> 
> Probably the most basic way of implementing it without breaking anything. Hopefully has a chance of getting merged.. De/serialization is only going to get heavier and heavier in the future.

How does it work btw? I don't understand that too well.

**Comment by bob7l** [comment]
> > > In branch [chat-optimization-v2](https://github.com/Janmm14/BungeeCord/tree/chat-optimization-v2) I have created an implementation of lazy deserialization in packets, with methods to hopefully retain full bytecode compatibility (Edit: for the individual packets, not for DefinedPacket methods).
> > > Edit 2: Its untested btw. Maybe someone has a better way of implementing it? It should be possible to replace the long generics of Deserializable with static types, for shorter code, but I'm unsure if thats desired over flexibility towards possible similar attempts for other complex deserialization.
> > 
> > 
> > Probably the most basic way of implementing it without breaking anything. Hopefully has a chance of getting merged.. De/serialization is only going to get heavier and heavier in the future.
> 
> How does it work btw? I don't understand that too well.

He's not deserializating the string payload unless it's requested. So when it comes time to write the packet, it just writes the string and doesn't need to serialize the text components back into a string assuming it never deserialized

**Comment by Janmm14** [comment]
> > > > In branch [chat-optimization-v2](https://github.com/Janmm14/BungeeCord/tree/chat-optimization-v2) I have created an implementation of lazy deserialization in packets, with methods to hopefully retain full bytecode compatibility (Edit: for the individual packets, not for DefinedPacket methods).
> > > > Edit 2: Its untested btw. Maybe someone has a better way of implementing it? It should be possible to replace the long generics of Deserializable with static types, for shorter code, but I'm unsure if thats desired over flexibility towards possible similar attempts for other complex deserialization.
> > > 
> > > 
> > > Probably the most basic way of implementing it without breaking anything. Hopefully has a chance of getting merged.. De/serialization is only going to get heavier and heavier in the future.
> > 
> > 
> > How does it work btw? I don't understand that too well.
> 
> He's not deserializating the string payload unless it's requested. So when it comes time to write the packet, it just writes the string and doesn't need to serialize the text components back into a string assuming it never deserialized

Once deserialized, the code will ignore the previously serialized string as the deserialized representation could have been changed/updated, because it was requested.

I think I am using the `Deserializable` thing just with the same generics parameters. Should I, before creating the pull request, switch it to be `ChatComponentHolder` or so?

**Comment by bob7l** [comment]
> I think I am using the `Deserializable` thing just with the same generics parameters. Should I, before creating the pull request, switch it to be `ChatComponentHolder` or so?

I'd just open the pull and see what md_5 says. I think Deserializable is perfect.

**Comment by Janmm14** [comment]
> I'm not sure I understand the question

Whether in chat-optimzation-v2 it should stay as `Deserializeable<String, BaseComponent>` or if I should remove the generics and call it sth like `ChatComponentHolder`

**Comment by md-5** [comment]
You could have a fixed subclass of the generic one? I don't think it really matters

**Comment by Janmm14** [comment]
> You could have a fixed subclass of the generic one? I don't think it really matters

That doesn't wok this easily, a fixed sub-interface would need different implementations. So I kept it like it was.
PR here: #3634

**Comment by NEZNAMY** [comment]
While on this topic I found something interesting I wanted to ask about if anyone knows.
BungeeCord deserializes scoreboard packets coming from backend in order to kick all players when a plugin accidentally double-registers a team, however, this deserialized packet is not forwarded into the channel pipeline like other packets (such as PlayerListItem), but instead, the raw bytebuf is. Why is that? Why does BungeeCord not forward the decoded scoreboard packet, but raw bytebuf instead? And why does it properly forward other packets?

**Comment by Janmm14** [comment]
@NEZNAMY please link or mention the involved places in code. I do not see this happening anywhere.

**Comment by NEZNAMY** [comment]
That's the issue - I didn't find the part of the code that is "wrong". While I disagree with what you are saying, that's not the point here. The point is that BungeeCord deserializes these packets, but when injecting custom duplex handler with a plugin and reading "write" direction, only the raw buffer is forwarded, so if I want to listen to those packets, I need to deserialize them once again, which overcomplicates things and slows down performance as well.

**Comment by andreasdc** [comment]
> That's the issue - I didn't find the part of the code that is "wrong". While I disagree with what you are saying, that's not the point here. The point is that BungeeCord deserializes these packets, but when injecting custom duplex handler with a plugin and reading "write" direction, only the raw buffer is forwarded, so if I want to listen to those packets, I need to deserialize them once again, which overcomplicates things and slows down performance as well.

Is this what you mean? https://github.com/SpigotMC/BungeeCord/issues/3503

**Comment by NEZNAMY** [comment]
That's one of the reasons why BungeeCord deserializes these packets, yes. Since the job is already done, would be nice to see the deserialized packet forwarded to the pipeline instead of the raw buffer. Or is it to save resources while encoding?

**Comment by Janmm14** [comment]
> That's the issue - I didn't find the part of the code that is "wrong". While I disagree with what you are saying, that's not the point here. The point is that BungeeCord deserializes these packets, but when injecting custom duplex handler with a plugin and reading "write" direction, only the raw buffer is forwarded, so if I want to listen to those packets, I need to deserialize them once again, which overcomplicates things and slows down performance as well.

Actually this is intended behaviour and in fact an optimization inside the ChannelWrapper's write method. It applies to all registered packets those handlers in the classes extending "PacketHandler" (like DownsteamBridge) are not throwing CancelSendSignal.INSTANCE.

**Comment by NEZNAMY** [comment]
> > That's the issue - I didn't find the part of the code that is "wrong". While I disagree with what you are saying, that's not the point here. The point is that BungeeCord deserializes these packets, but when injecting custom duplex handler with a plugin and reading "write" direction, only the raw buffer is forwarded, so if I want to listen to those packets, I need to deserialize them once again, which overcomplicates things and slows down performance as well.
> 
> Actually this is intended behaviour and in fact an optimization inside the ChannelWrapper's write method. It applies to all registered packets those handlers in the classes extending "PacketHandler" (like DownsteamBridge) are not throwing CancelSendSignal.INSTANCE.
> 
> Either way, your code is not using API and some further bungeecord optimizations that are laying dormant in PRs here, but implemented for example in [IvanCord](https://github.com/MrIvanPlays/IvanCord/tree/master) could break your system of ignoring all existing bungeecord tooling for packet reading and hacking your way into the netty pipeline. Move the code to your spigot servers and edit the packets right where they are created! Keep the bungee clean, stable and running fast.

Thanks for the info, looks like it is intended.
Moving code to backend would make things drastically more complicated than they are on proxy installation and since I'm doing public development, tons of things are out of my control, making it even harder and having to do these kinds of things to at least partially gain control over important things. Public development sucks. I should leave.
